system/hardware/tici/tests/test_power_draw.py

```diff
@@ -48,35 +50,75 @@ def setUp(self):
   def tearDown(self):
     manager_cleanup()
 
+  def get_expected_messages(self, proc):
+    return int(sum(SAMPLE_TIME * SERVICE_LIST[msg].frequency for msg in proc.msgs))
+
+  def tabulate_msg_counts(self, msgs_and_power):
+    msg_counts = defaultdict(lambda: 0)
+    for _, counts in msgs_and_power:
+      for msg, count in counts.items():
+        msg_counts[msg] += count
+    return msg_counts
+
+  def get_power_with_warmup_for_target(self, proc, prev):
+    socks = {msg: messaging.sub_sock(msg) for msg in proc.msgs}
+    for sock in socks.values():
+      messaging.drain_sock_raw(sock)
+
+    msgs_and_power = deque([None] * SAMPLE_TIME, maxlen=SAMPLE_TIME)
+
+    start_time = time.time()
+
+    while (time.time() - start_time) < MAX_WARMUP_TIME:
+      power = get_power(1)
+      iteration_msg_counts = {}
+      for msg,sock in socks.items():
+        iteration_msg_counts[msg] = len(messaging.drain_sock_raw(sock))
+      msgs_and_power.append((power, iteration_msg_counts))
+
+      if any(m is None for m in msgs_and_power):
```

Comment:
I think checking the length is more clear here

Reply:
done